### PR TITLE
Support runtime Topic designation

### PR DESCRIPTION
**Pull request by rnarubin**
Previously the Topic type was only constructible from static strings. This works for code-gen messages, but doesn't work if the topic needs to be determined at runtime. This change permits constructing Topics from arbitrary strings

### CODE REVIEW

**Comment by rnarubin** [comment]
@blogle

File: src/topic.rs [hunk 1]
```diff
@@ -1,21 +1,36 @@
+use smallstr::SmallString;
+
 /// A message queue topic name to which messages can be published
-#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
-pub struct Topic(&'static str);
+// A survey of common topics found lengths between 16 and 35 bytes
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Topic(SmallString<[u8; 36]>);
```
Comment (blogle) [inline src/topic.rs:6]: I did run into this problem, since I am creating topic names dynamically - but worked around it by leaking the strings and then using a hashmap to cache the mapping from (site id, prio) -> &'static str

File: src/backends/googlepubsub/publisher.rs [hunk 2]
```diff
@@ -374,16 +374,21 @@ where
...
-                                    .into_project_topic_name(client.project()),
+
+                    // avoid cloning the topic if the key exists
+                    match this.topic_sinks.get_mut(&topic) {
+                        Some(existing) => existing,
+                        None => this.topic_sinks.entry(topic.clone()).or_insert(Box::pin(
+                            TopicSink::new(
+                                client.client.publish_topic_sink(
+                                    TopicName::new(topic.as_ref())
+                                        .into_project_topic_name(client.project()),
```
Comment (blogle) [inline src/backends/googlepubsub/publisher.rs:385]: This particular line is what I think needs to change for me to send messages across various projects. Though you might be wanting to address that in a separate PR.

File: src/topic.rs (hunk 1 above)
Comment (blogle) [inline src/topic.rs:6]: is this saying we can support any string up to 36Bytes? I wonder if any of our topics exceed that? Or is this similar to a stackvec and will falback to heap allocation if we exceed that threshold?

Comment (rnarubin) [reply src/topic.rs:6]: It's backed by SmallVec, which yeah does the fallback to heap

File: src/backends/googlepubsub/publisher.rs (hunk 2 above)
Comment (rnarubin) [reply src/backends/googlepubsub/publisher.rs:385]: yes, different PR

File: src/topic.rs (hunk 1 above)
Comment (rnarubin) [reply src/topic.rs:6]: I guess there's a reasonable bound on that combination, so that's not the worst. But we can do better!

**Review by blogle — APPROVED** [review]
